botorch/generation/gen.py

```diff
@@ -219,11 +219,12 @@ def gen_candidates_torch(
         if verbose:
             print("Iter: {} - Value: {:.3f}".format(i, -(loss.item())))
         loss_trajectory.append(loss.item())
-        param_trajectory["candidates"].append(candidates.clone())
+        param_trajectory["candidates"].append(candidates.clone().detach())
 
         def closure():
             bayes_optimizer.zero_grad()
-            loss.backward()
+            output_grad = torch.autograd.grad(loss, candidates)[0]
+            candidates.grad = output_grad
```

Comment:
hmm there seems to be some inconsistency here between the unconstrained candidates, the clamped candidates, and those were `fixed_features` is applied. Specifically, `bayes_optimizer` uses `clamped_candidates` as its parameter (l. 207), but here we're taking the gradient w.r.t. `candidates`. There also seems to be some inconsistency below where we compute `clamped_candidates` but return `candidates` (and evaluate the acquisition function on those). 

We probably want to combine the clamping and the "fixing" of features into a single wrapper so we can always work with a single `projected_candidates` tensor to avoid this kind of confusion.